File: apps/portal/src/lib/api/schemas/elementInspections.ts
import { z } from 'zod';

import {
  BorgingsmomentPhaseEnum,
  BorgingsmomentStatusEnum,
  ChecklistItemSchema,
} from './borgingsplan';
import { ChecklistItemResultSchema } from './inspection';

export const ElementInspectionItemSchema = z.object({
  checklist_item: ChecklistItemSchema,
  result: z.union([ChecklistItemResultSchema, z.null()]),
  moment_name: z.string(),
  moment_phase: BorgingsmomentPhaseEnum,
  moment_status: BorgingsmomentStatusEnum,
});
export type ElementInspectionItem = z.infer<typeof ElementInspectionItemSchema>;

export const ElementInspectionsResponseSchema = z.object({
  items: z.array(ElementInspectionItemSchema),
  element_global_id: z.string(),
  file_id: z.string().uuid(),
});
export type ElementInspectionsResponse = z.infer<typeof ElementInspectionsResponseSchema>;

File: apps/api/src/bimstitch_api/schemas/element_inspections.py
"""Response schemas for the element-inspections endpoint (backlog #49).

Maps an IFC element (by ``global_id`` + ``file_id``) to the checklist items
linked to that element, together with their inspection results (if any).
"""

from uuid import UUID

from pydantic import BaseModel, ConfigDict

from bimstitch_api.models.borgingsmoment import BorgingsmomentPhase, BorgingsmomentStatus
from bimstitch_api.schemas.borgingsplan import ChecklistItemRead
from bimstitch_api.schemas.inspection import ChecklistItemResultRead


class ElementInspectionItem(BaseModel):
    """One checklist item linked to the queried IFC element, together with
    its parent moment context and optional inspection result."""

    model_config = ConfigDict(from_attributes=True)

    checklist_item: ChecklistItemRead
    result: ChecklistItemResultRead | None
    moment_name: str
    moment_phase: BorgingsmomentPhase
    moment_status: BorgingsmomentStatus


class ElementInspectionsResponse(BaseModel):
    """Aggregate response grouping all checklist items linked to a single IFC
    element in a specific file."""

    items: list[ElementInspectionItem]
    element_global_id: str
    file_id: UUID
